File: knittingapp.client/src/types/Schema.ts

export interface Loop {
    m: number;
    n: number;
    color: string;
    type: string;
}


export interface Schema {
    schemaId?: string;
    schemaName: string;
    loopMap: LoopMap;
    schemaImage?: string;
}

export class LoopMap {
    m: number;
    n: number;
    loopMap: Loop[][];
    loopWidth?: number;
    loopHeight?: number;
    nullM?: number;
    nullN?: number;
    constructor(m: number, n: number, loopWidth?: number, loopHeight?: number, nullM?:number, nullN?:number) {
        this.m = m;
        this.n = n;
        this.loopHeight = loopHeight;
        this.loopWidth = loopWidth;
        this.nullM = nullM;
        this.nullN = nullN;

        // Создаём матрицу с петлями по умолчанию
        this.loopMap = Array(m).fill(null).map((_, x) =>
            Array(n).fill(null).map((_, y) => ({
                m: x,           // можно использовать координаты
                n: y,
                color: "#FFFFFF",  // белый цвет по умолчанию
                type: "loop"           // тип loop по умолчанию
            }))
        );
    }

    // Получить значение из ячейки
    getLoop(x: number, y: number): Loop | null {
        if (x >= 0 && x < this.m && y >= 0 && y < this.n) {
            return this.loopMap[x][y];
        }
        return null;
    }

    // Обновить значение в ячейке
    changeLoopColor(x: number, y: number, color:string): void {
        if (x >= 0 && x < this.m && y >= 0 && y < this.n) {
            this.loopMap[x][y].color = color;
        }
    }

    changeLoopType(x: number, y: number, type: string): void {
        if (x >= 0 && x < this.m && y >= 0 && y < this.n) {
            this.loopMap[x][y].type = type;
        }
    }
}

File: KnittingApp/Schema.cs
using static KnittingApp.SharedConstants;
namespace KnittingApp
{
    public class Schema
    {
        public Guid userId { get; }
        public Guid schemaId { get; set; }
        public int m;  //ширина
        public int n;  //высота
        public string schemaName { get; set; }
        public Guid LoopMapId { get; }

        public LoopMap loopMap { get; set; }
        public string schemaImage { get; set; }

        public Schema(Guid loopMapId, Guid userId)
        {
            LoopMapId = loopMapId;
            this.userId = userId;
        }
        public Schema(int m, int n, string schemaName, Guid userId)
        {
            this.m = m;
            this.n = n;
            this.schemaName = schemaName;
            loopMap = new LoopMap(m, n);
            LoopMapId=loopMap.LoopMapId;
            loopMap.LoopsReaderId = null;
            schemaId = Guid.NewGuid();
            schemaImage = "null";
            this.userId=userId;
        }

        public void ChangeLoopColor(int m, int n, string color)
        {
            if (loopMap == null)
            {
                throw new NullReferenceException("LoopMap равно null");
            }
            loopMap.ChangeColor(m, n, color);
        }

        public void ChangeLoopType(int m, int n, LoopType type)
        {
            if (loopMap == null)
            {
                throw new NullReferenceException("LoopMap равно null");
            }
            loopMap.ChangeType(m, n, type);
        }

        public void ColorSchema(string color)
        {
            loopMap.ColorLoopMap(color);
        }

    }
}
